### PR TITLE
237 middleware per procedure

### LINKED ISSUE #237: Middleware per procedure
Currently we can add middleware to a runtime service (gateway, node, proxy) that is applied to each procedure request. Next we want to be able to specify what middleware to apply to what procedure.

For this we need to change the registration of middleware a bit, by naming them in the starter file:

```ts
startServer(moduleImporter).then(server =>
{
    // Before
    server.addMiddleware(new AuthMiddleware());

    // After
    server.registerMiddleware('auth', new AuthMiddleware());
});
```

We also need to extend the segment configuration:

```json
{
    "./account/update.js": {
        "default": {
            "access": "public",
            "middlewares": ["auth"]
        }
    }
}
```

To make it easier to add the middleware to a group of procedures, we want to also be able to configure them on module and segment level. An example configuration could look like this:

```json
{
    "middlewares": ["logging"],
    "modules": {
        "./account/update.js": {
            "middlewares": ["auth"]
            "procedures": {
                "default": {
                    "access": "public",
                    "middlewares": ["notify"]
                }
            }
        }
    }
}
```

**Note** that the implementation will break existing segment configurations because additional configuration levels have been added. We can prevent this by using __special__ keys beginning with a specific specifier (e.g. ~) like this:

```json
{
    "~middlewares": ["logging"],
    "./account/update.js": {
        "~middlewares": ["auth"]
        "default": {
            "access": "public",
            "middlewares": ["notify"]
        }
    }
}
```

This isn't the most beautiful option, but it doesn't break existing configurations, is easier and quicker to configure. We need to decide what configuration style to use before picking up this issue.

Next, we want to make the middleware configurable from the segment file to prevent registering a lot of specific in the starter file like this:

```ts
startServer(moduleImporter).then(server =>
{
    // Avoid this situation
    server.registerMiddleware('auth-admin-department-x', new AuthMiddleware('admin', 'department-x'));
    server.registerMiddleware('auth-admin-department-y', new AuthMiddleware('admin', 'department-y'));
    server.registerMiddleware('auth-manager-department-x', new AuthMiddleware('manager', 'department-x'));
    server.registerMiddleware('auth-manager-department-y', new AuthMiddleware('manager', 'department-y'));
    server.registerMiddleware('auth-customer-type-a', new AuthMiddleware('customer', 'type-a'));
    /* ... */
});
```

In the desired situation we only register the middleware once, and provide the construction arguments from the configuration like this:

```json
{
    "./some/module.js": {
        "doMagic": {
            "access": "public",
            "middlewares": {
                "auth": { "role": "manager", "scope": "department-x" }
            }
        }
    }
}
```

For this we need to figure out how we want to create new middleware instances from the configuration. The easiest is to only register the middleware class in the starter file like this:

```ts
startServer(moduleImporter).then(server =>
{
    // Before (register an instance)
    server.registerMiddleware('auth', new AuthMiddleware());

    // After (register the class)
    server.registerMiddleware('auth', AuthMiddleware);
});
```

The `AuthMiddleware` should specify the configured properties as constructor parameters:

```ts
export default class AuthMiddleware implements Middleware
{
    constructor(role: string, scope: string) { /* store values */ }

    /* handle function */
}
```

Our reflection library can be used for checking the configured arguments and the creation of the instance.

### PR DESCRIPTION
**Pull request by petermasking**
Fixes #237

Changes proposed in this pull request:
- Improved middleware implementation (request / response models)
- Updated middleware documentation
- Added migration information

@MaskingTechnology/jitar

### CODE REVIEW

File: packages/runtime/src/models/Response.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        this.#headers = headers;
+    }
+
+    get result() { return this.#result; }
+
+    set result(value: unknown) { this.#result = value; }
+
+    get headers() { return this.#headers; }
+
+    clearHeaders()
```
Comment (BasMasking) [inline packages/runtime/src/models/Response.ts:19]: The return type is missing.

File: packages/runtime/src/models/Response.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        this.#headers.clear();
+    }
+    
+    setHeader(name: string, value: string)
```
Comment (BasMasking) [inline packages/runtime/src/models/Response.ts:24]: The return type is missing.

File: packages/runtime/src/models/Response.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        this.#headers.set(name, value);
+    }
+
+    getHeader(name: string): string | undefined
+    {
+        return this.#headers.get(name);
+    }
+
+    removeHeader(name: string)
```
Comment (BasMasking) [inline packages/runtime/src/models/Response.ts:34]: The return type is missing.

File: packages/runtime/src/models/Request.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        this.#headers.set(name, value);
+    }
+
+    getHeader(name: string): string | undefined
+    {
+        return this.#headers.get(name);
+    }
+
+    removeHeader(name: string)
```
Comment (BasMasking) [inline packages/runtime/src/models/Request.ts:57]: The return type is missing.

File: packages/runtime/src/models/Request.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        this.#args.delete(name);
+    }
+
+    clearHeaders()
+    {
+        this.#headers.clear();
+    }
+
+    setHeader(name: string, value: string)
```
Comment (BasMasking) [inline packages/runtime/src/models/Request.ts:47]: The return type is missing.

File: packages/runtime/src/models/Request.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        return this.#args.get(name);
+    }
+
+    removeArgument(name: string)
+    {
+        this.#args.delete(name);
+    }
+
+    clearHeaders()
```
Comment (BasMasking) [inline packages/runtime/src/models/Request.ts:42]: The return type is missing.

File: packages/runtime/src/models/Request.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        this.#args.set(name, value);
+    }
+
+    getArgument(name: string): unknown
+    {
+        return this.#args.get(name);
+    }
+
+    removeArgument(name: string)
```
Comment (BasMasking) [inline packages/runtime/src/models/Request.ts:37]: The return type is missing.

File: packages/runtime/src/models/Request.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    get fqn() { return this.#fqn; }
+
+    get version() { return this.#version; }
+
+    get args() { return this.#args; }
+
+    get headers() { return this.#headers; }
+
+    setArgument(name: string, value: unknown)
```
Comment (BasMasking) [inline packages/runtime/src/models/Request.ts:27]: The return type is missing.

File: packages/runtime/test/services/LocalNode.spec.ts [hunk 9]
```diff
@@ -52,28 +53,32 @@ describe('services/LocalNode', () =>
...
 
-            expect(result).toBe('fourth');
+            expect(response.result).toBe('fourth');
         });
 
         it('should not run a non-existing procedure', async () =>
         {
-            const run = async () => node.run('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const request = new Request('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const run = async () => node.run(request);
```
Comment (BasMasking) [inline packages/runtime/test/services/LocalNode.spec.ts:81]: I expect this name to be `response` to be consistent with the other tests

File: packages/runtime/test/services/NodeBalancer.spec.ts [hunk 10]
```diff
@@ -31,7 +32,8 @@ describe('services/LocalGateway', () =>
     {
         it('should throw a node not available error', async () =>
         {
-            const run = async () => emptyBalancer.run('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const request = new Request('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const run = async () => emptyBalancer.run(request);
```
Comment (BasMasking) [inline packages/runtime/test/services/NodeBalancer.spec.ts:36]: I expect this name to be `response` to be consistent with the other tests

File: packages/runtime/test/services/LocalGateway.spec.ts [hunk 11]
```diff
@@ -55,21 +56,24 @@ describe('services/LocalGateway', () =>
...
 
-            expect(result).toBe('fourth');
+            expect(response.result).toBe('fourth');
         });
 
         it('should not run a non-existing procedure', async () =>
         {
-            const run = async () => gateway.run('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const request = new Request('nonExisting', Version.DEFAULT, new Map(), new Map());
+            const run = async () => gateway.run(request);
```
Comment (BasMasking) [inline packages/runtime/test/services/LocalGateway.spec.ts:76]: I expect this name to be `response` to be consistent with the other tests

File: packages/runtime/test/services/LocalNode.spec.ts (hunk 9 above)
Comment (petermasking) [reply packages/runtime/test/services/LocalNode.spec.ts:81]: Same question here as in the `LocalGateway` comment.

File: packages/runtime/test/services/NodeBalancer.spec.ts (hunk 10 above)
Comment (petermasking) [reply packages/runtime/test/services/NodeBalancer.spec.ts:36]: Same question here as in the LocalGateway comment.

File: packages/runtime/test/services/LocalGateway.spec.ts (hunk 11 above)
Comment (BasMasking) [reply packages/runtime/test/services/LocalGateway.spec.ts:76]: You're right about the different approach, as it indeed contains an anonymous run function. I'll close the comments as Resolved

**Review by BasMasking — APPROVED** [review]

**Pull request merged**